server/server.go

```diff
@@ -6,53 +6,299 @@ package server
 			}
-			break
+
+		case <-producer.Context().Done():
+			return fmt.Errorf("producer context: %w", producer.Context().Err())
+
+		case <-serv.polling.ctx.Done():
+			return fmt.Errorf("server context: %w", serv.polling.ctx.Err())
 		}
-		res := messages.EventResult{EventId: evt.EventId, Timestamp: pbts.Now()}
-		results = append(results, &res)
 	}
-	return &messages.PublishReply{Results: results}, nil
 }
 
-// StreamAcknowledgements is the server implementation of the gRPC StreamAcknowledgements call
-func (serv shipperServer) StreamAcknowledgements(streamReq *messages.StreamAcksRequest, prd pb.Producer_StreamAcknowledgementsServer) error {
+// Close implements the Closer interface
+func (serv *shipperServer) Close() error {
+	serv.close.Do(func() {
+		// polling must be stopped first, otherwise it would try to write
+		// a notification to a closed channel and this would cause a panic
+		serv.polling.stop()
+		<-serv.polling.stopped
+
+		serv.logger.Debug("shutting down all notifications...")
+		serv.notifications.shutdown()
+		serv.logger.Debug("all notifications have been shut down")
+	})
+
+	return nil
+}
+
+func (serv *shipperServer) startPolling() {
+	go func() {
+		ticker := time.NewTicker(serv.cfg.PollingInterval)
+		defer ticker.Stop()
 
-	// we have no outputs now, so just send a single dummy event
-	evt := messages.StreamAcksReply{Acks: []*messages.Acknowledgement{{Timestamp: pbts.Now(), EventId: streamReq.Source.GetInputId()}}}
-	err := prd.Send(&evt)
+		for {
+			select {
 
+			case <-serv.polling.ctx.Done():
+				err := serv.polling.ctx.Err()
+				if err != nil && errors.Is(err, context.Canceled) {
+					serv.logger.Error(err)
+				}
+				close(serv.polling.stopped) // signaling back to `Close`
+				return
+
+			case <-ticker.C:
+				serv.logger.Debug("updating indices...")
+				err := serv.updateIndices(serv.polling.ctx)
+				if err != nil {
+					serv.logger.Errorf("failed to update indices: %s", err)
+				} else {
+					serv.logger.Debug("successfully updated indices.")
+				}
+			}
+		}
+	}()
+}
+
+// updateIndices updates in-memory indices and notifies subscribers if necessary.
+// TODO: this is a temporary implementation until the queue supports the `persisted_index`.
+func (serv *shipperServer) updateIndices(ctx context.Context) error {
+
+	// TODO: for now we calculate an approximate value
+	// we cannot rely on this value and needs to be changed when the queue is ready.
+	metrics, err := serv.queue.Metrics()
 	if err != nil {
-		return fmt.Errorf("error in StreamAcknowledgements: %w", err)
+		return fmt.Errorf("failed to fetch queue metrics: %w", err)
 	}
+
+	log := serv.logger
+	c := change{}
+
+	if metrics.UnackedConsumedEvents.Exists() {
+		oldPersistedIndex := serv.GetPersistedIndex()
+		persistedIndex := serv.GetAcceptedIndex() - int64(metrics.UnackedConsumedEvents.ValueOr(0))
+
+		if persistedIndex != oldPersistedIndex {
+			atomic.SwapInt64(&serv.persistedIndex, persistedIndex)
```

Comment:
```go
		if persistedIndex != oldPersistedIndex {
			atomic.SwapInt64(&serv.persistedIndex, persistedIndex)
```

I'm not sure if this code is going to stay after the persisted index support is added to the queue, but I think there is a concurrency bug here. The `persistedIndex` can change between when the value was read originally and when it is swapped here. 

At minimum you would probably want to use https://pkg.go.dev/sync/atomic#CompareAndSwapInt64 instead and maybe handle the case where the value conflict.

It would maybe be slower to just wrap these values in a mutex, but it would be easier to avoid concurrency issues than it is with atomics.

Reply:
> The persistedIndex can change between when the value was read originally and when it is swapped here.

This is not a problem here because:

1. the persisted index will be way slower in change anyway (the final output has to acknowledge the events, it takes time), so the value does not have to be the most recent
2. this value is not supposed to be contextual to the current request, it's just an update that is sent with each response for convenience instead of subscribing via `PersistedIndex` endpoint. Normally if clients are interested in this counter they should use a `PersistedIndex` in a background go routine that would ack their event batches once `persisted_index` >= `accepted_index` (from the `PublishReply`).

Reply:
My concern would be that API clients could observe `persistedIndex` as non-monotonic based on the ordering of the swaps here, where that would never happen if they were using the `PersistedIndex` call itself.

Even if it is for convenience, it should not have unexpected behaviour. That makes it significantly less useful and I'd be worry about difficult to debug issues in clients depending on what they use this for (which we can't completely control).

Reply:
I'll replace it with a CAS then.

Reply:
I just realised that I cannot use CAS for that because it swaps only if the old value is not equal the value under the given address, not the other way around.

I still don't understand the concern though, can you give me an example when this code causes a concurrency issue?

If I remove that `if` check it would be bulletproof, this `if` is just an optimisation, not more. Keep in mind that this function is the only place where the `serv.persistedIndex` value is changed.